test/gem_test.rb

```diff
@@ -4,29 +4,57 @@
 
 class GemTest < ProtoBoeuf::Test
   def test_can_be_required
-    Tempfile.create do |file|
-      file.write(<<~RUBY)
-        require "bundler/inline"
+    assert_rb_runs_successfully(<<~RUBY, "Failed to require the gem")
+      require "bundler/inline"
+
+      gemfile do
+        gem "protoboeuf", path: "#{__dir__}/..", require: true
+      end
+
+      ::ProtoBoeuf
+
+      # The following should autoloaded
+      ::ProtoBoeuf::CodeGen
+      ::ProtoBoeuf::Google::Api::FieldBehavior
+      ::ProtoBoeuf::Google::Protobuf::Any
+      ::ProtoBoeuf::Google::Protobuf::FileDescriptorProto
+      ::ProtoBoeuf::Google::Protobuf::FileDescriptorSet
+
+      exit 0
+    RUBY
+  end
+
+  def test_codegen_can_be_required
+    # Test that we fixed a circular require bug: https://github.com/Shopify/protoboeuf/issues/173
 
-        gemfile do
-          gem "protoboeuf", path: "#{__dir__}/..", require: true
-        end
+    _, stderr, _ = assert_rb_runs_successfully(<<~RUBY)
+      require "bundler/inline"
 
-        ::ProtoBoeuf
+      gemfile do
+        gem "protoboeuf", path: "#{__dir__}/..", require: false
+      end
 
-        # The following should autoloaded
-        ::ProtoBoeuf::CodeGen
-        ::ProtoBoeuf::Google::Api::FieldBehavior
-        ::ProtoBoeuf::Google::Protobuf::Any
-        ::ProtoBoeuf::Google::Protobuf::FileDescriptorProto
-        ::ProtoBoeuf::Google::Protobuf::FileDescriptorSet
+      require "protoboeuf"
+      require "protoboeuf/codegen"
 
-        exit 0
-      RUBY
+      exit 0
+    RUBY
 
+    refute_match(/warning: loading in progress, circular require considered harmful/, stderr)
```

Comment:
I like being specific, but I'm a bit worried that any change in the warning message could cause this test to pass unexpectedly. These messages aren't really stable across Ruby versions. I'm not sure if JRuby or TruffleRuby even emit them identically.